sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ShouldImplementExportedInterfacesBase.cs

```diff
@@ -0,0 +1,126 @@
+            new HashSet<KnownType>
+            {
+                KnownType.System_ComponentModel_Composition_ExportAttribute,
+                KnownType.System_ComponentModel_Composition_InheritedExportAttribute,
+            };
+    }
+
+    public abstract class ShouldImplementExportedInterfacesBase<TArgumentSyntax, TExpressionSyntax, TClassSyntax>
+        : ShouldImplementExportedInterfacesBase
+        where TArgumentSyntax : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+        where TClassSyntax : SyntaxNode
+    {
+        protected static bool IsOfExportType(ITypeSymbol type, INamedTypeSymbol exportedType) =>
+            type.GetSelfAndBaseTypes()
+                .Union(type.AllInterfaces)
+                .Any(currentType => currentType.Equals(exportedType));
+
+        protected INamedTypeSymbol GetExportedTypeSymbol(SeparatedSyntaxList<TArgumentSyntax>? attributeArguments,
+            SemanticModel semanticModel)
+        {
+            if (attributeArguments == null)
+            {
+                return null;
+            }
+
+            var arguments = attributeArguments.Value;
+            if (arguments.Count == 0 ||
+                arguments.Count > 2)
+            {
+                return null;
+            }
+
+            var argumentSyntax = GetArgumentFromNamedArgument(arguments) ??
+                GetArgumentFromSingleArgumentAttribute(arguments) ??
+                GetArgumentFromDoubleArgumentAttribute(arguments, semanticModel);
+            var typeOfExpression = GetExpression(argumentSyntax);
+
+            var exportedTypeSyntax = GetExportedTypeSyntax(typeOfExpression);
+            if (exportedTypeSyntax == null)
+            {
+                return null;
+            }
+
+            return semanticModel.GetSymbolInfo(exportedTypeSyntax).Symbol as INamedTypeSymbol;
+        }
+
+        private TArgumentSyntax GetArgumentFromNamedArgument(IEnumerable<TArgumentSyntax> arguments) =>
+            arguments.FirstOrDefault(x => GetIdentifier(x) == "contractType");
+
+        private TArgumentSyntax GetArgumentFromSingleArgumentAttribute(IEnumerable<TArgumentSyntax> arguments)
+        {
+            if (arguments.ElementAtOrDefault(1) != null)
+            {
+                return null;
+            }
+
+            // Only one argument, should be typeof expression
+            return arguments.ElementAtOrDefault(0);
+        }
+
+        private TArgumentSyntax GetArgumentFromDoubleArgumentAttribute(IEnumerable<TArgumentSyntax> arguments,
+            SemanticModel semanticModel)
+        {
+            var firstArgument = GetExpression(arguments.ElementAtOrDefault(0));
+            if (firstArgument != null &&
+                semanticModel.GetConstantValue(firstArgument).Value is string)
+            {
+                // Two arguments, second should be typeof expression
+                return arguments.ElementAtOrDefault(1);
```

Comment:
We are enumerating several times this IEnumerable (agreed, it's not that large). wouldn't it be simple to pss the argument as `SeparatedSyntaxList<TArgumentSyntax>` and use the indexer directly?

Reply:
It would be more complex because we would have to test the size before doing the access.